Android Auto

puff opened the pull request:
Adds basic Android Auto and Android Automotive support.

## Features:
* Browse & play albums, playlists, and albums in each genre.
* Artists start an instant mix when playing them.
* Jump to albums/artists by letter (click A-Z on scroll bar).
* Seek forwards and backwards in songs.
* Skip to previous and next songs in queue, as well as directly to any song in the queue.
* Shuffle button
* Offline mode support.
  * Even outside of offline mode, it will try to use downloaded items before going online, unless it is fetching all items for root category (e.g. albums, artists, etc).
* Search (voice and keyboard)

#### Future features/TODO?
* Repeat button
* Favorite button
* Instant mix button
* Home tab
* Genres should start an instant mix (not yet implemented in Finamp)
* Offline mode artists/genres instant mix (not yet implemented in Finamp)

Currently, there is a hard limit of ~~100~~ 250 (online mode) or 1000 (offline mode) items displayed so the app doesn't crash on large libraries.

#### // dev notes
* Investigate why it doesn't switch to the Now Playing screen sometimes when starting playback, or takes a while before switching to it. This is likely because of the playback states being wrong.
* Add pagination so we can show more than 100 items. [This](https://developer.android.com/reference/android/support/v4/media/MediaBrowserCompat#EXTRA_PAGE()) and [This](https://proandroiddev.com/how-to-paginate-mediabrowserservice-e4fdc902ff2) may be useful.

jmshrv commented:
This is awesome to see finally :)

Is the 100 item limit just an awkward thing with Android Auto or a thing that specifically needs to be fixed here? Jellyfin luckily makes pagination very easy when fetching from the server.

Artwork sounds like an annoying issue to fix though 🙃

Chaphasilor commented:
Thanks for the PR! I can try to take a look at it in the coming days, I just need to find a car to test with ^^

I'm guessing there is no built-in pagination support in Android Auto, and I don't think it would be super useful anyway.  
I believe focussing on quick actions, like shuffle all songs, instant mixes and playing recently played items is a good approach.

Search and some voice interactions would definitely be useful down the road (hah), but if basic playback and browsing is working then I guess it qualifies for a beta.  
Maybe it attracts some additional contributors :P

puff commented:
>I just need to find a car to test with ^^

You can use the emulator! https://github.com/jmshrv/finamp/issues/24#issuecomment-1734229067 It's very convenient for debugging too.

> I'm guessing there is no built-in pagination support in Android Auto, and I don't think it would be super useful anyway. I believe focussing on quick actions, like shuffle all songs, instant mixes and playing recently played items is a good approach.

Yeah, I think I agree with you. Not sure how useful pagination would be while in the car.

> Search and some voice interactions would definitely be useful down the road (hah)

Yeah, 100%. I wanted to incorporate that into this PR, but I was having an issue I didn't really understand where it crashes when searching.

Chaphasilor commented:
Yeah I know about the emulator and have used it last time, but in the end I want to make sure everything works on an actual car too ^^

Regarding voice, how flexible is that? Can it only be used for voice search, or could we enable custom actions, like saying "Start playing Hello by Adele" and it will start an instant mix?

puff commented:
> Can it only be used for voice search, or could we enable custom actions, like saying "Start playing Hello by Adele" and it will start an instant mix?

I know you can do stuff like that with [app actions](https://developer.android.com/guide/app-actions/overview), though I have no idea how yet. I can't even get Assistant to recognize Finamp's name lol

Chaphasilor commented:
The second link you provided seems to be geared towards navigation/Points-of-Interest apps. Here's the one for media apps: https://developer.android.com/training/cars/media#support_voice

Seems like it has everything we need. If the users indicates an item type (album, playlist, etc.) we can use that to deal with Jellyfin's poor search, and if they omit the type "Play some music" we could resort to shuffle all for now. Would you like to look into it? :)

puff commented:
> Would you like to look into it? :)
Sure, I have some time the next few days so I'll see if I can understand this and get something usable.

Chaphasilor commented:
Awesome. Let me know if you need help with anything!

puff commented:
Thanks for the feedback, I agree with your observations here.

> For albums and playlists I'd really like to either be prompted if I want to play in-order or shuffled
I don't think this is possible, but a setting for the default would work.

> The sorting is a bit different than in the phone app
> The jump-to-letter function is neat, but only works with the loaded items

I don't think I can do anything about these :(, they seem to be built into Android Auto.

> There's also a "for you" section...

Seems like [this](https://developer.android.com/media/implement/assistant/driving#media_recommendations_in_driving_mode) is related to that section. 

> There's a display bug when shuffle is active

I've done a little troubleshooting, and I'm confused. It seems there may be a bug somewhere causing a conflict between just_audio and the new queue service?

https://github.com/jmshrv/finamp/blob/2ea2156bcbbe78d844997046f7b310539fef53de/lib/services/music_player_background_task.dart#L377

The queue service expects `queueIndex` to be the non-shuffled, original index:
https://github.com/jmshrv/finamp/blob/2ea2156bcbbe78d844997046f7b310539fef53de/lib/services/queue_service.dart#L94-L101
https://github.com/jmshrv/finamp/blob/2ea2156bcbbe78d844997046f7b310539fef53de/lib/services/queue_service.dart#L148-L151

I suspect somewhere in just_audio expects it to be the shuffled index and is sending it to the backend (ExoPlayer?), which results in the queue showing the wrong song as selected. To elaborate, the erroneous selected song in queue is in the position of the original (non-shuffled) index, which is why I suspect this.

Changing `queueIndex` to the shuffled index, and replacing both occurrences of `adjustedQueueIndex` in queue_service directly to `_queueAudioSourceIndex` fixes the issue, however I'm not sure of the implications of this.

Chaphasilor commented:
We could reduce the number of tabs even further, with a home tab and a library tab? On the library tab we could have a grid with the different categories.

Genre mixes should be possible, at least there's a way to add a genre to a mix on the phone 🤔 But maybe that just adds all individual albums, not sure.

Can we change the sorting within Android Auto? If something like "Recently Added" is possible, then it should be possible to use the BaseItemDto's SortTitle too.

I'll take a look at the modifications to the queue service. just_audio is really opinionated and the layered architecture doesn't make things easier. I've struggled with it for many monhs 🙃

Chaphasilor commented:
@puff I managed to fix up the queue based on your proposed changes. There were a few other things I needed to adjust, and I fixed a bug that I introduced a while ago (items in next up where not being reported to the external/OS queue).  
From my testing everything seems to work as it should, but please let me know if anything is off!  
With the beta release approaching fast I would hate to break things, and there have been the strangest bugs with playback and queue management in the past...

puff commented:
> We could reduce the number of tabs even further, with a home tab and a library tab?

Yeah, I like that idea. 

> Genre mixes should be possible

I didn't see a way already implemented in Finamp yet, but yeah it should be possible if the Jellyfin app has it.

> Can we change the sorting within Android Auto?

Yes, we can sort the MediaItem list before returning from `getChildren`. While testing, I noticed the offline sort already used doesn't match output from the Jellyfin api.
https://github.com/jmshrv/finamp/blob/4463bff0998f22c4558f1178d9d90c399527f090/lib/components/MusicScreen/music_screen_tab_view.dart#L327

Looks to be because the Jellyfin api is case-insensitive:
https://github.com/jellyfin/jellyfin/blob/7027bc00fc06314929011735e425763779cb4076/MediaBrowser.Controller/Entities/BaseItem.cs#L897

I've fixed it and added sort to Android Auto.

Chaphasilor commented:
Thanks for fixing that. There are some sorting-related issues that need to be tackled at some point anyway, like #444 or #581, so if you feel like taking a look while you're dealing with sorting anyway, that would be much appreciated.  
Otherwise I'll try to fix it after some bigger issues have been fixed...

puff commented:
I managed to get artwork to work with offline items and (kind've) Android Automotive using the [android_content_provider](https://pub.dev/packages/android_content_provider) package.

The white icon in the middle is the placeholder art.
![image](https://github.com/jmshrv/finamp/assets/33184334/47cf2cbb-d304-4989-be99-08d980fed2cd)

Only problem is that for Android Automotive, we can't use `http(s)` URIs. Because of this, we have to resolve them ourselves. My current implementation hangs until every item has a resolved image. I think the way to fix this is to offload it to the ContentProvider. For now, I've commented it out until I have a better solution.

Chaphasilor commented:
Did you manage to take a look at the voice command stuff? Or is there anything that you need help with? :)

puff commented:
> Did you manage to take a look at the voice command stuff? Or is there anything that you need help with? :)

I've been having some trouble with it. It doesn't default to finamp even when it's open (maybe I forgot to configure something though?). It really doesn't like to recognize the app's name, you have to say it slowly. Saying "album" or really anything else other than the name of the song/album makes it default to YouTube Music (probably your default music service) so extras is always null.

"play some music on finamp"
![image](https://github.com/jmshrv/finamp/assets/33184334/825cdea6-2c71-4313-957d-cdea28c576cb)
Seems like some generic phrases are stripped out, odd that the "on finamp" part wasn't though.

"play black ray on finamp"
![image](https://github.com/jmshrv/finamp/assets/33184334/2b9b906e-86c2-43c3-b8a5-ff8c4bc1f2c6)

If you want to try it out, you just need to add this to the manifest:
```xml
<intent-filter>
  <action android:name="android.media.action.MEDIA_PLAY_FROM_SEARCH" />
  <category android:name="android.intent.category.DEFAULT" />
</intent-filter>
```
and uncomment `playFromSearch` in `music_player_background_task.dart`

Chaphasilor commented:
Sorry for not responding earlier, I was a bit busy. Just tested it out and didn't manage to invoke the `playFromSearch()` method even once, although I did manage to get it to recognize the app's name a few times. I did manage to open the app by saying "Play some music on Finamp", but I don't think that's related to the manifest changes.  
If the problem would just be to properly recognize the app name, then [this might help](https://stackoverflow.com/questions/61606687/google-assistant-training/61616766#61616766), if it still works. But I think the whole voice command stuff seems to be pretty flawed, with YouTube being the default for a lot of things even after explicitly not selecting a default in the Assistant settings.  
I only tested with Google Assistant, not with Android Auto, but I'd expect results to stay the same.

I guess we should skip this feature for now, including regular search. Search is due for a rewrite soon, and it makes little sense to integrate the old version now...

puff commented:
Thanks for trying it. As for the link you provided, it looks like the app may need to be published on Play Store in order to access app actions and change the pronunciation.

I changed the default music service in the Assistant app's settings to "No default provider" and it now defaults to finamp instead of YouTube Music when you say "play album x" or "play some music". This isn't really ideal though because users would probably have to do this too.
The extras is null because audio_service doesn't actually pass it to our callback. Someone made a PR for it here: https://github.com/ryanheise/audio_service/pull/1015
Might be nice to have extras because Google sometimes provides us the album and artist name.
![image](https://github.com/jmshrv/finamp/assets/33184334/42cae737-f771-4b90-a14e-7cfafa79ee5a)
(Note for future: don't rely on this 100%, it should search for query if it cant find it with given artist/album name)

Also, I found out that Android Auto apps need to support voice commands to be published on Play Store.

Chaphasilor commented:
They have to support voice commands? o.O  
Could you share a link to the relevant docs? Is there a specific type they need to support, and does it have to go through the Google Assistant, or would speech-to-text for search be enough?

Requiring voice commands, which currently aren't really working, kinda sucks. I guess we could just throw it in the manifest and pretend like we support them, but I'm not sure.  
What are your thoughts on this? Would you prefer merging some form of this quickly, or do you want to spend more time on this and try to figure it out fully? :)

puff commented:
> Could you share a link to the relevant docs?

https://developer.android.com/training/cars/media#support_voice
https://developer.android.com/docs/quality-guidelines/car-app-quality#app_categories

This and some github issues suggest they require voice commands.
https://github.com/signalapp/Signal-Android/issues/6124
https://github.com/ryanheise/audio_service/issues/956#issuecomment-1635113977

> I guess we could just throw it in the manifest and pretend like we support them, but I'm not sure

I don't think so, from the above github issues it seems they test them.

Maybe we could disable Android Auto (comment it in manifest) in the Play Store release until we have it implemented? I would like to get it working though so we don't have to... I'll try implementing the search algorithm for it.

Chaphasilor commented:
Okay, got it. I still can't get a simple "Play [song title]" to work on Finamp. It still defaults to YouTube on my phone (although it fails to actually play), and appending "on Finamp" at the end makes it open YouTube and actually start playing, super strange. Did you try it through Android Auto or also just with the regular Assistant?  
As long as we can get a simple "Play [song title]" command to work, we should be fine.

Here's my quick implementation of how the search could work, it starts an instant mix for the first match returned by Jellyfin's search:

```dart
  @override
  Future<void> playFromSearch(String query, [Map<String, dynamic>? extras]) async {
    _audioServiceBackgroundTaskLogger.info("playFromSearch: $query ; extras: $extras");

    final jellyfinApiHelper = GetIt.instance<JellyfinApiHelper>();
    final finampUserHelper = GetIt.instance<FinampUserHelper>();
    final audioServiceHelper = GetIt.instance<AudioServiceHelper>();

    try {
      
      final searchResult = await jellyfinApiHelper.getItems(
        parentItem: finampUserHelper.currentUser?.currentView,
        includeItemTypes: "Audio",
        searchTerm: query.trim(),
        isGenres: false,
        startIndex: 0,
        limit: 1,
      );

      if (searchResult!.isEmpty) {
        return;
      }

      _audioServiceBackgroundTaskLogger.info("Playing from search query: ${searchResult[0].name}");
      await audioServiceHelper.startInstantMixForItem(searchResult[0]);

    } catch (err) {
      _audioServiceBackgroundTaskLogger.severe("Error while playing from search query:", err);
    }

    return;
  }
```
It's untested, since I could never get it to be called...

puff commented:
> Did you try it through Android Auto or also just with the regular Assistant?'

Android Auto. I tried with the Assistant app on phone and I get the same issue as you. Using the [media controller test app (MCT)](https://github.com/googlesamples/android-media-controller) allows you to call `playFromSearch` directly though.

When using voice commands, try also saying "song" or "album" before the name like "play song/album x."

Your search code does work for songs. However, the test in the MCT app sometimes fails. Probably something with async stuff. I'm not sure if this matters though, because the song does actually play.

Chaphasilor commented:
Would you mind sending me an APK for the test app? I couldn't find a ready-made one and am not super familiar with Android Studio :)

I'll try your tips and take a look at the search issues. I think it's expected to use different states while preparing playback, and that is probably missing in our case. Not sure if there's anything we can do about it though...

puff commented:
Sure, here's a compiled apk for the MCT app [mediacontroller-release.zip](https://github.com/jmshrv/finamp/files/14157138/mediacontroller-release.zip)
I used IntelliJ IDEA to compile it.

Chaphasilor commented:
After playing around with this, I can confirm that playing from search works. In Android Auto it works by default ("play song xyz"). I also observed that Google Assistant seems to "enhance" search queries in some cases: "play 404" results in a response "okay, asking FinAmp to play 404 by Knife Party". I'm pretty sure the artist name is not coming from Finamp in this case, but I could be wrong.

I found no way to make the tests pass by fixing the state. There are some functions within audio_service that should be working, but don't really do anything, for example setting the queue title. Since we have a queue source, it would be nice to show it, but adding a value to `queueTitle` within the BaseAudioHandler doesn't work.

I also tried implementing the regular search after reading that [we can show search results that the user can choose from in case the wrong item was picked](https://developer.android.com/training/cars/media#display_search), which isn't uncommon with Jellyfin's crappy search. I managed to get the search results to show, but selecting a result doesn't actually start playback yet. I'm working on fixing that though :)

Chaphasilor commented:
Okay, managed to get search and voice search alternative results to work: 

https://github.com/jmshrv/finamp/assets/18015147/858086e1-bc2a-447e-85db-7f365a8d8b56

I guess the only thing that we need to do now is try to get any many of the Android Auto tests to pass as possible. The main issues here seem to be "skipToNext"/"skipToPrevious"/"skipToQueueItem", and the media artwork test. Not sure if browse tree depth and custom action icon type are actually required.  
I might replace the custom action icons at some point, so that should fix one of the tests.

One thing I'd like you to take a look at is improving the handling of the `mediaId`s. I know it sucks that we are only passed the `mediaId` and nothing else, but maybe it's possible to get the required into about the "category" from somewhere else instead, like the context / active tab (or lack thereof)? If not, I'd still like to replace the "-1" (and now "-2") special IDs with some kind of enum that can be stringified and reliably compared, and comes with additional meaning. Maybe instead of splitting the ID you could also simply use a JSON string with custom fields instead, to "properly" pass all the required data? Would be great if you could take a look at it :)

It really seems like we're slowly but surely getting there! :D

Chaphasilor commented:
@puff so the PR you mentioned go merged yesterday and I just played around with it. We get the extras now, and I implemented playing back songs, albums and artists now. I think that's all the types that we can differentiate, playlists don't seem to work.  
Albums are played in-order (even if I say "shuffle album xyz..." we don't get any extras provided to indicate that) and artists play as instant mixes.  

Song and album results are also (rather crudely) filtered by trying to match the artist provided by Google Assistant to the search results. This does work surprisingly well, and is kind of necessary because whenever GA knows a song, it replaces the query with "<song name> <artist name>", breaking Jellyfin's search and making the "Search results" button mostly useless.  
The problem here is that while `playFromSearch()` receives the proper extras, `search()` does not, so we cannot tell what the title and artist are. So I tried to get the best match possible right from the initial voice search by using the provided metadata.  
Take a look at the commit message for the necessary syntax for requesting different items.

Chaphasilor commented:
@puff are you still around? I can handle the refactoring, but is there anything else that needs to be done before this can be merged and released? 😅

puff commented:
Sorry for the (very) late reply!
> maybe it's possible to get the required into about the "category" from somewhere else instead

I don't think it is. An enum sounds like a good idea though.

> try to get any many of the Android Auto tests to pass as possible

I think these only fail because of the playback state not being set in time because the functionality works.

> is there anything else that needs to be done 

The only thing I know of is the bug with artwork on Android Automotive (not Auto), it doesn't display artwork for online items.
Pagination can be added later, I don't think it's necessary to be able to scroll more than 100 items. Maybe a configurable limit in settings though?

Also, fantastic work on the search features!

Chaphasilor commented:
Awesome, and thank you so much! <3  

I'll try to get the enum in then.  
Could you take a look at the tests and see if your theory is correct? I'm a bit paranoid that the failing tests might cause the app to be rejected during Play Store review...

Chaphasilor commented:
@puff I replaced the mediaId stuff with a serializable class now, a simple enum wasn't really enough.  
While trying it out I also found out there's `AudioService.recentRootId` which we might be able to use to provide content for the "For You" section.  

If you could give the changes a quick test to make sure I didn't break anything, that would be nice. I had some troubles with external IDs at first, and I'm not sure I caught everything...

puff commented:
> If you could give the changes a quick test

I did a quick test and it broke everything except search :P
Playing music (albums, artists) and browsing genres are broken.

About the "For you" page, I don't think it's possible without PR'ing [EXTRA_SUGGESTED](https://developer.android.com/reference/androidx/media/MediaBrowserServiceCompat.BrowserRoot#EXTRA_SUGGESTED()) into audio_service. [This](https://developer.android.com/media/implement/assistant/driving#media_recommendations_in_driving_mode) says it is passed into `onGetRoot` as a root hint like how `AudioService.recentRootId` is.
[EXTRA_OFFLINE](https://developer.android.com/reference/android/service/media/MediaBrowserService.BrowserRoot#EXTRA_OFFLINE) sounds interesting too.

puff commented:
I don't really know why the tests are failing. I found a PR for audio_service that supposedly fixed it, but it got rejected. In the replies, the author of it comments his app got approved without the PR so I assume that means they don't need to pass.
https://github.com/ryanheise/audio_service/pull/1042#issuecomment-1722520668

Chaphasilor commented:
Okay, we can just go ahead and try then, it'll probably be hit-or-miss with the reviewing anyway.  
I did manage to fix the issues for non-search items. If you have the time, please give it another try. I'll get this merged tomorrow if everything is working :)

puff commented:
Looks like everything is working now! :D

Chaphasilor commented on lib/services/android_auto_helper.dart:
```diff
@@ -35,22 +35,19 @@ class AndroidAutoHelper {
   }
 
   Future<List<BaseItemDto>> getBaseItems(MediaItemId itemId) async {
-    final tabContentType = TabContentType.values.firstWhere((e) => e == itemId.contentType);
```
Oh wow, yeah that was a bit redundant 😂

Chaphasilor commented:
Are you also getting freezes sometimes when starting playback for a item? It produces logs like these, and the app stops responding:

```log
I/sd.finamp.debug(29917): Thread[6,tid=29927,WaitingInMainSignalCatcherLoop,Thread*=0xb400007209038200,peer=0x12c02d88,"Signal Catcher"]: reacting to signal 3
I/sd.finamp.debug(29917): 
I/sd.finamp.debug(29917): Wrote stack traces to tombstoned
I/sd.finamp.debug(29917): Thread[6,tid=29927,WaitingInMainSignalCatcherLoop,Thread*=0xb400007209038200,peer=0x12c02d88,"Signal Catcher"]: reacting to signal 3
I/sd.finamp.debug(29917): 
I/sd.finamp.debug(29917): Wrote stack traces to tombstoned
I/flutter (29917): [QueueService/FINEST] 2024-02-17 12:30:42.854113: Saved new periodic queue previous:[] current:328e57c8369e334101424468b4b6a50f seek:62685 next:[] queue:[]
I/sd.finamp.debug(29917): Thread[6,tid=29927,WaitingInMainSignalCatcherLoop,Thread*=0xb400007209038200,peer=0x12c02d88,"Signal Catcher"]: reacting to signal 3
I/sd.finamp.debug(29917): 
I/sd.finamp.debug(29917): Wrote stack traces to tombstoned
```

puff commented:
I've reproduced it, it's hanging in `AudioService.internalStartForeground` -> `AudioService.buildNotification` (java).
Because of this, I think it has something to do with battery optimization. 
Can you try adding `android:foregroundServiceType="mediaPlayback"` to the service tag in the manifest here? 
https://github.com/jmshrv/finamp/blob/redesign/android/app/src/main/AndroidManifest.xml#L32
Since doing this, I'm unable to reproduce it again, but I might just be getting lucky.

Chaphasilor commented:
Interesting. I only saw the problem (if it's the same one) when starting playback for an album or artist, and only after getting rid of the string-based IDs.  
Please keep me posted. We can merge this once you give your OK :)

Chaphasilor commented:
Could it be that the calls are recursive? [The docs](https://developer.android.com/reference/android/content/ContentProvider#onCallingPackageChanged()) mention the following:

> This typically happens when a [`ContentProvider`](https://developer.android.com/reference/android/content/ContentProvider) makes a nested call back into itself when already processing a call from a remote process.

It could also be a bug in `android_content_provider`.  
Unless it's happening to you for basic interaction, I think we can still get this merged. I've marked it as experimental in the release notes anyway :)

puff commented:
> Unless it's happening to you for basic interaction

It hasn't happened during normal behavior yet. Did you do anything special to trigger it?

I did notice another bug though. If you play a song during startup before the queue service loads the previous queue, it can cause odd behavior. Seems like a simple fix though, just need to check if queue state is loading or not.

Chaphasilor commented:
Did nothing special, music was playing and I clicked on an album. I'll try to repro it later. Maybe it was related to a loading queue?

Chaphasilor commented:
I actually just got a "For You" recommendation:  
![image](https://github.com/jmshrv/finamp/assets/18015147/dcae8140-a2a7-4881-808d-e67f99f13d65)

I did listen to that playlist yesterday, but the app data was cleared completely before I started Android Auto. Maybe the OS is remembering it through the media session stuff, or are we somehow providing that data? 🤔

Chaphasilor commented:
Here's another log dump. I have the feeling it the freezing gets more likely the less time there is in-between playback requests, but I haven't done enough testing to confirm that.  
I also noticed some timeouts, and increased the timeout duration slightly. Maybe the content-provider requests are timing out (does it make http requests?) and the timeout isn't handled properly?

```log
I/AudioTrack(11045): set(): streamType -1, sampleRate 44100, format 0x1, channelMask 0x3, frameCount 14144, flags 0, notificationFrames 0, sessionId 62057, transferType 3, uid 10483, pid 11045
W/XingSeeker(11045): XING data size mismatch: 6023927, 6023799
I/flutter (11045): [PlaybackHistoryService/FINE] 2024-02-19 20:50:00.067900: Reporting seek event for Peacock
I/flutter (11045): [PlaybackHistoryService/INFO] 2024-02-19 20:50:00.068289: Starting playback progress for Peacock
W/System  (11045): A resource failed to call close. 
W/XingSeeker(11045): XING data size mismatch: 5195166, 5195038
D/BufferPoolAccessor2.0(11045): bufferpool2 0xb4000070891e70e8 : 5(40960 size) total buffers - 4(32768 size) used buffers - 17/22 (recycle/alloc) - 6/34 (fetch/transfer)
D/BufferPoolAccessor2.0(11045): bufferpool2 0xb40000708918fe78 : 0(0 size) total buffers - 0(0 size) used buffers - 2186/2191 (recycle/alloc) - 6/4370 (fetch/transfer)
D/BufferPoolAccessor2.0(11045): evictor expired: 2, evicted: 1
D/BufferPoolAccessor2.0(11045): evictor expired: 1, evicted: 1
I/sd.finamp.debug(11045): Thread[6,tid=11052,WaitingInMainSignalCatcherLoop,Thread*=0xb400007209038200,peer=0x12c00ef0,"Signal Catcher"]: reacting to signal 3
I/sd.finamp.debug(11045): 
I/sd.finamp.debug(11045): Wrote stack traces to tombstoned
F/crash_dump64(16325): crash_dump.cpp:500] failed to attach to thread 1116: Permission denied
I/sd.finamp.debug(11045): Thread[6,tid=11052,WaitingInMainSignalCatcherLoop,Thread*=0xb400007209038200,peer=0x12c00ef0,"Signal Catcher"]: reacting to signal 3
I/sd.finamp.debug(11045): 
I/sd.finamp.debug(11045): Wrote stack traces to tombstoned
F/crash_dump64(16353): crash_dump.cpp:500] failed to attach to thread 1005: Permission denied
```

Chaphasilor commented:
@puff do you prefer we get this over quickly and ship it with the initial release, or to make sure everything is fully working and ship it with a later update? Everything else is ready for the release, we just need to finalize the code for the release :)

puff commented:
> Were you able to gain any insights?

I'm still not sure why the calls aren't completing inside `android_content_provider`. An overview of the behavior happening is like this:
- User plays item 1.
- `audio_service` calls `setMediaItem`, which blocks the background thread and `AudioService` class in java because it's synchronized. 
- This completes normally.
- User plays item 2.
- `audio_service` again calls `setMediaItem`. 
- During this, `audio_service` receives a callback from the OS for building the notification (`buildNotification()`) for item 1. 
  - This blocks the main thread for `audio_service` as it has to wait for the `setMediaItem` call to complete.
  - However, `setMediaItem` never completes because the `android_content_provider` hangs while waiting for a response from our content provider. I don't know why we don't ever get a response. I ***think*** that maybe `audio_service` is blocking the flutter thread while waiting for a response from `setMediaItem`, causing a deadlock where `android_content_provider` can't send back a response to the java method.

I know it's disappointing, but I think we should wait until this is fixed before its merged. If this bug happened to me while driving, it'd be frustrating and possibly distracting.

Chaphasilor commented:
Okay, thanks for looking into this. I agree that we should wait for now, but it won't be that long anyway.  
Just out of curiosity, if we left out the android_content_provider package, would it still work on Android Auto? I remember you mentioning that this package fixes something with Android Automotive, but I could be wrong of course!

puff commented:
> Okay, thanks for looking into this. I agree that we should wait for now, but it won't be that long anyway. Just out of curiosity, if we left out the android_content_provider package, would it still work on Android Auto? I remember you mentioning that this package fixes something with Android Automotive, but I could be wrong of course!

We need it to display offline artwork in Android Auto and offline + online artwork for Android Automotive. Here's an explanation on why: https://github.com/ryanheise/audio_service/issues/1049#issuecomment-1810294899

puff commented:
I think this is an issue with audio_service, so I've made an issue there: https://github.com/ryanheise/audio_service/issues/1063
This issue is a little confusing to me though, so I'm gonna wait for a response there to do anything more.

Chaphasilor commented:
Okay, thanks for looking into it so thoroughly! I'll hopefully be able to take care of the merge conflict in the coming days as well :)

Chaphasilor commented:
So that means everything should be working now? 😁  
If so, I should be able to take a look at the merge conflicts tomorrow, and maybe we can push it with the next update!

puff commented:
Yeah, everything should be working now.

Chaphasilor commented:
Sweet! I tested it an so far everything seems to be working great. Not a single crash or freeze yet.  
I also made some more improvements to search, like using metadata/extras from previous voice searches to search and prioritize matches when using the "Search results" button on the player screen. Finamp now also searches playlists *and* songs if the item type is unspecified ("play xyz", "play song xyz", or "play playlist xyz" all have the same extras without metadata), and if a matching playlist is found, that is prioritized over matching songs. If the user wanted to play a song with that name instead, they can then use the "Search results" button.  
So everything should be relatively flexible now.

I also finished up the merge, currently testing that.

Chaphasilor commented:
Would be nice if you could do another round of testing, focusing especially on the offline stuff you implemented, since I tried to migrate everything to the new system :)  
I'll also try to test, and then we can merge it (maybe tomorrow?)

puff commented:
Had to fix `includeItemTypes` in `getBaseItems()`, but now it looks like everything is working again.

Chaphasilor commented on lib/services/android_auto_helper.dart:
```diff
@@ -51,6 +51,11 @@ class AndroidAutoHelper {
   }
 
   Future<List<BaseItemDto>> getBaseItems(MediaItemId itemId) async {
+    // offline mode only supports albums and playlists for now (no offline instant mix for others yet)
+    if (FinampSettingsHelper.finampSettings.isOffline && (itemId.contentType == TabContentType.artists || itemId.contentType == TabContentType.genres)) {
+      return [];
+    }
+
```
That is not correct any more. New download system supports albums, playlists and artists. Basically, all collections.  
Instant mixes are a different beast, but playback of artists and genres should be possible (genre playback isn't implemented yet, but that is true for online mode as well

Chaphasilor replied:
I managed to implement this for artists, where I just add all downloaded songs per album to queue. I'll try and see if I can also get genres to work, since it's just browsing downloaded albums I no reason why it wouldn't be possible.

puff replied:
Good catch!

Chaphasilor commented:
I noticed that when you tried to use downloaded items by default even if online, it resulted in Finamp only playing downloaded songs from i.e. albums, even though we could just fetch the other songs from the server.  
Since the stream URL is automatically resolved from downloads if available when creating a `MediaItem` for the queue, I changed this so Finamp always prefers the server info.  

Maybe we could fall back to only downloaded items in case fetching the items from the server fails to due a spotty connection, but I'm not sure if we want to do that in general. We don't do it in the regular app after all, but maybe driving justifies the changed behavior? What do you think?

Chaphasilor commented:
@puff do you know of any way to get the currently selected tab in such a way that we could use it do decide what to search for?  
E.g. if on the albums tab and clicking the search icon, albums will be searched, but on the artists tab artists. 

I'm asking because there are many users that listen to albums, and there currently is no good way to pick albums that are not in the first 100.

puff commented:
> Maybe we could fall back to only downloaded items...

I think we should just retry fetching it until we get a response, or change the timeout to be longer.

> do you know of any way to get the currently selected tab 

We could store it during the `getChildren()` call. If it's a root collection, it should be our current tab.

Chaphasilor commented:
I also noticed a lot of these errors in the logs:

```log
I/flutter ( 5291): Error loading artUri: Invalid argument(s): Unsupported scheme 'content' in URI content://com.unicornsonlsd.finamp/data/user/0/com.unicornsonlsd.finamp.debug/files/images/7a1c3729c8866e731912b055ccef98d5.jpg
I/flutter ( 5291): #0      _HttpClient._openUrl (dart:_http/http_impl.dart:2756:9)
I/flutter ( 5291): #1      _HttpClient.openUrl (dart:_http/http_impl.dart:2623:7)
I/flutter ( 5291): #2      IOClient.send (package:http/src/io_client.dart:117:38)
I/flutter ( 5291): #3      HttpFileService.get (package:flutter_cache_manager/src/web/file_service.dart:37:44)
I/flutter ( 5291): #4      WebHelper._download (package:flutter_cache_manager/src/web/web_helper.dart:115:24)
I/flutter ( 5291): #5      WebHelper._updateFile (package:flutter_cache_manager/src/web/web_helper.dart:97:28)
I/flutter ( 5291): <asynchronous suspension>
I/flutter ( 5291): #6      WebHelper._downloadOrAddToQueue (package:flutter_cache_manager/src/web/web_helper.dart:65:7)
I/flutter ( 5291): <asynchronous suspension>
```

Any idea what's going on? I suspect it's something to do with the content provider...

Chaphasilor commented:
I'll also try to tackle offline voice search tomorrow. That doesn't work at all yet, but should be doable.

puff commented:
> I also noticed a lot of these errors in the logs: ...
> Any idea what's going on? I suspect it's something to do with the content provider...

Seems like an incompatibility with our content provider and audio_service's `preloadArtwork` option. The `_loadArtwork` method used with preloading artwork in the queue doesn't handle `content://` URIs. They are only handled when the current media item is being loaded. It'd probably require a small refactor (on our side too for efficiency) to add support for preloading content:// artworks in audio_service, so I think we should just disable `preloadArtwork`.

> but there are also cases where we can enter search (i.e. from the player screen), where it's not clear which tab should be used. 

I think it should just prioritize certain types in the results.
Playlists -> Artists -> Albums -> Songs is what Jellyfin orders by in their app when searching, so I think it would make sense here too. `includeItemTypes` is comma-delimited so it should be easy to search all types (except for genres, since Jellyfin doesn't search those).

Chaphasilor commented:
Ah, yeah I enabled that option recently because I was hoping it would avoid the artist sometimes not loading correctly in the Android notification. I'll disable it for now.

I'll see what I can do about the search.

Chaphasilor commented:
It's been way to long since I last worked on this, but I'm planning to get this into 0.9.8 or 0.9.9 (depending on how quickly we push out a "hotfix" for the things that broke in 0.9.7).  

I guess what's missing right now is the search improvements (search different types + show headers for the different types) and some fixes I wanted to add for voice commands in offline mode, if I remember correctly.

Chaphasilor commented:
@puff could it be that http-only URLs don't work with the content provider? I tried testing while using the local IP of my server, and the images only showed up for downloaded media. Logging in over the public HTTPS address everything worked normally 🤔

puff commented:
It works for my local server over http. Using a Pixel 8 and the DHU emulator. Also works on my Android Automotive emulator.
Maybe there's something in your logs?

Chaphasilor commented:
Nothing in the logs sadly. Only logs I get are `Getting children of <xyz>` and `Couldn't get offline image [...]` (for the ones that don't have an image, where the placeholder is correctly being applied).

Here's what it looks like:

![image](https://github.com/jmshrv/finamp/assets/18015147/9d99f03f-8df9-4bf6-b4ef-133c54809e6a)
![image](https://github.com/jmshrv/finamp/assets/18015147/f0823593-ffdd-43b3-8065-2e27cd69206b)
![image](https://github.com/jmshrv/finamp/assets/18015147/843152af-e701-4da2-881b-45b95dd33bbd)
![image](https://github.com/jmshrv/finamp/assets/18015147/4db49d2b-de8e-41b1-9be5-6dbb30f89ee3)

I can try testing it with an actual car in the coming days, maybe even tomorrow.

Chaphasilor commented:
Thanks for the review Komodo! The age of this PR is starting to show...

I'll try to go through your comments later and improve the sections as best as I can :)

puff commented on lib/services/queue_service.dart:
```diff
@@ -973,18 +976,43 @@ class QueueService {
+
+    // replace with content uri or jellyfin api uri
+    if (downloadedImage != null) {
+      artUri = downloadedImage.file?.uri.replace(scheme: "content", host: "com.unicornsonlsd.finamp");
```
Good catch, should probably have a platform check here.

puff commented on lib/main.dart:
```diff
@@ -111,6 +112,21 @@ void main() async {
 
+    // Load the album image from assets and save it to the documents directory for use in Android Auto
+    final documentsDirectory = await getApplicationDocumentsDirectory();
+    final albumImageFile = File('${documentsDirectory.absolute.path}/images/album_white.png');
```
`/images/album_white.png` could be moved to a constant. Not really sure what else to do here.

Komodo5197 replied:
I basically mean using the path_helper to combine the segments.  Like I said, I'm not sure how much it matters though.

Chaphasilor commented on lib/components/MusicScreen/music_screen_tab_view.dart:
```diff
@@ -57,6 +57,7 @@ class _MusicScreenTabViewState extends State<MusicScreenTabView>
 
   final _jellyfinApiHelper = GetIt.instance<JellyfinApiHelper>();
   final _isarDownloader = GetIt.instance<DownloadsService>();
+  final _finampUserHelper = GetIt.instance<FinampUserHelper>();
```
Not anymore, at least according to the linter ^^

Chaphasilor commented on lib/models/finamp_models.dart:
```diff
@@ -619,6 +619,24 @@ enum TabContentType {
     }
   }
+
+  static TabContentType fromItemType(String itemType) {
```
It is, here:  
https://github.com/puff/finamp/blob/289f3b27288a8c4e08985e87dab496def6c1dfdd/lib/services/android_auto_helper.dart#L697

Should I be using something else?

Komodo5197 replied:
I'm pretty sure this was based off some other code that got factored out completely.  This seems fine, although I'm wondering if using BaseItemDto types would make more sense than tab types for all this?  I guess it is mapping to basically music screen tabs though.

Komodo5197 replied:
I'm pretty sure this was based off some other code that got factored out completely. This seems fine, although I'm wondering if using BaseItemDto types would make more sense than tab types for all this? I guess it is mapping to basically music screen tabs though.

Chaphasilor replied:
Yeah that is the case, it really is closer to the tab types (so the browsable collections containing the items)

Chaphasilor commented:
Also before I forget, the way Symfonium handles playing in-order vs. shuffled is by making each "leave" item a node/collection, which has two children, "Play in order" and "Play shuffled". I think that would also be a good solution for us, and we could even add more choices, like starting and instant mix or adding to Next Up.

Chaphasilor commented on lib/services/music_player_background_task.dart:
```diff
@@ -429,6 +440,127 @@ class MusicPlayerBackgroundTask extends BaseAudioHandler {
+
+  // triggers when skipping to specific item in android auto queue
+  @override
+  Future<void> skipToQueueItem(int index) async {
```
I think `skipToIndex()` is clearer, and it doesn't really hurt to do it this way?

Komodo5197 replied:
I guess not?  It's pretty strange to have this pointless redirection though.

Chaphasilor replied:
I actually don't think so. There are some weird external methods that the OS can decide to call, but unifying this into a more sensible API seems like a good idea to me.  
Basically, all we're doing is say "if Android Auto decides it needs to skip to a specific queue item, that just means we skip to the provided index, just like we do in some other cases". That redirection is basically some documentation/explanation saying that the two actions are equivalent, without there being any confusion...

Komodo5197 commented on lib/services/android_auto_helper.dart:
```diff
@@ -0,0 +1,761 @@
+    // fetch the online version if we can't get offline version
+
+    // select the item type that each parent holds
+    final includeItemTypes = itemId.parentType == MediaItemParentType.collection // if we are browsing a root library. e.g. browsing the list of all albums or artists
```
This should be a switch expression for all the cases checking the same field.

Chaphasilor commented on lib/services/android_auto_helper.dart:
```diff
@@ -0,0 +1,761 @@
+        || tabContentType == TabContentType.artists || tabContentType == TabContentType.songs;
+  }
+
+  Future<MediaItem> _convertToMediaItem({
```
Like this?

Chaphasilor commented on lib/services/music_player_background_task.dart:
```diff
@@ -429,6 +440,127 @@ class MusicPlayerBackgroundTask extends BaseAudioHandler {
+  }
+
+  // menus
+  @override
```
Is it clear enough now? Did I miss an important method?

Chaphasilor commented:
Okay, aside from breaking the whole Android Auto media browser, the review findings should be handled now 😅  
I'll try to figure out what exactly broke tomorrow...

Chaphasilor commented:
There we go. Managed to fix it today :D

Chaphasilor commented:
Okay, the enhanced search is also working now. I'd like to get a like button onto the AA player screen / media notification because it's been a long-requested feature, and then this should be ready to merge

Chaphasilor commented:
> Also before I forget, the way Symfonium handles playing in-order vs. shuffled is by making each "leaf" item a node/collection, which has two children, "Play in order" and "Play shuffled". I think that would also be a good solution for us, and we could even add more choices, like starting and instant mix or adding to Next Up.

Custom browse actions (https://developer.android.com/training/cars/media#custom_browse_actions) would be a much better solution though...

Chaphasilor commented:
This should be ready to merge now. All the basics are working reasonably well. If someone could confirm that this still works fine on iOS that would also be nice :)

Chaphasilor commented:
Okay, I'll merge this in now for the next update, since most things should be working fine. If not, we won't know without having users test it anyway.  
Enhancements and new features can come at a later point, based on user feedback.

Chaphasilor commented:
Thank you for getting this started looking into the documentation, etc. @puff!

Chaphasilor commented:
@puff do you have a working Automotive setup? I'm low on disk space, but it seem like we need screenshots from Automotive for our Play Store listing to pass the review.

If you could grab some screenshots in landscape and portrait mode using the official demo server, that would be great!

frandj99 commented:
Hey guys, any update on this?
I'm really hoping to see a Jellifin client workin on Android Automotive and I just found this discussion.
May I ask if there's hope to see Finamp on car's Play Store soon ? 😅
This would be a game-changer for me (and I think also for other AAOS users)

Titaniumtown commented:
@frandj99 this was merged 5 months ago and is in the redesign branch. You can install the beta using the instructions in the README.md.

Chaphasilor commented:
@frandj99 while this has indeed been merged into the beta for some time, I'm not sure if it actually works with Android Automotive. Please try it out and let us know!

frandj99 commented:
@Titaniumtown @Chaphasilor I was finally able to build the app (also I had to go through a private test on Google Play Console in order to install on my car, cuz sideloading is not allowed 🥲)
I have some bad news so... I feel a login screen is missing. If I open the app I only get a message with "Loading contents..." as in this photo (sorry, it's in Italian)
![image](https://github.com/user-attachments/assets/b2263cb9-58f3-4a6d-8737-09fcd056fbb8)

I'm down to test anything, let me know if I'm missing something or so...

Chaphasilor commented:
@frandj99 is that Android Automotive? Does it install the app directly onto the car? 🤔

frandj99 commented:
> @frandj99 is that Android Automotive? Does it install the app directly onto the car? 🤔

Yes, it's Android Automotive of my Volvo XC40.
Let's say that installing any app I want is not easy 😅
Basically Android Automotive has its own Play Store application section, so I can download just a small number of apps approved for cars.

Sideloading is not allowed at all, I can't access the developer setting on Volvo AAOS "for security reasons", so the only way you can install any application is through the Play Store.
So what have I done? I paid Google for app publishing (even if I won't really do this), compiled the app with a different package name, added the string so I can use it while driving, create the signed app bundle and finally do the closed beta test with myself only, in order to see the app on my Play Store.

Yeah, looong and boring way, I needed some help by a friend of mine to compile the app in Android studio but finally today I was able to see the Finamp icon on my car 😄 (but it doesn't work)

Chaphasilor commented:
@frandj99 thanks for the info. I sadly current don't have time to work on this, but if you want to play around with it to get it somewhat working, you could open a (draft) PR, and we can merge it if there are any improvements to how it works at the moment!

richanshah commented:
in android auto there is page limit , I am getting java binder exception , has anyone given pagiantion support for same?

Chaphasilor commented:
@richanshah could you please clarify where you are getting the exception? Maybe record a video of it?
